### PR TITLE
Explicitly register dependent applications for modules

### PR DESCRIPTION
**Pull request by MengnanLi91**
This is a patch PR to explicitly register dependent applications. It is needed for the new [Applications] block I am working on (#23741). After parsing the application types from [Applications] block, this allows us to compare the parsed app type to application registration list and determine whether the correct application input is used. 

## Reason

It is needed to support [Applications] block working properly

## Impact

Support new capability

### CODE REVIEW

**Comment by loganharbour** [comment]
@permcody I thought that this was supposed to be satisfied by the module loader? I thought that we wanted to only enable modules in one place - the corresponding Makefile

**Comment by GiudGiud** [comment]
I think this has pretty deep implications for the creation of a new app. It's no longer a trivial process of just adding a yes in the Makefile. I can see this becoming a burden on user support

**Comment by loganharbour** [comment]
> I think this has pretty deep implications for the creation of a new app. It's no longer a trivial process of just adding a yes in the Makefile. I can see this becoming a burden on user support

It's not a requirement

**Comment by GiudGiud** [comment]
Ah ok it s just to get the Application block to work as expected? Or also the app_type block in MultiApps?

**Comment by MengnanLi91** [comment]
> Ah ok it s just to get the Application block to work as expected? Or also the app_type block in MultiApps?

This will provide the App register list we can and compare with the type input of Application block. It should work for MultiApps too

**Comment by permcody** [comment]
Looks like Logan and Guillaume settled on the right answer. This change shouldn't impact anything with new or existing apps. It simply allows us to create additional "app types" for nearly every app (including the modules). Right now only BlueCRAB and Direwolf register additional applications that a user may request (and as we know that's currently limited to only the Multiapp block). This PR allows somebody in a module to be able to explicitly create a depdendent app in a Multiapp.

As we move forward, we'll be able to explicitly create that same dependent application as _the_ top-level application for a given input. After this gets merged, we should patch the applications so applications have this same consistent interface.

**Review by loganharbour — CHANGES_REQUESTED** [review]
You'll need to change the coverage in `.coverage` for `scalar_transport` from 90 to 85 to satisfy https://civet.inl.gov/job/1946896/

File: .coverage [hunk 1]
```diff
@@ -91,3 +91,6 @@ require_total = 89
 
 [xfem]
 require_total = 80
+
```
Comment (loganharbour) [inline .coverage:94]: Not alphabetical?! come on... :D

**Review by loganharbour — APPROVED** [review]

**Review by loganharbour — APPROVED** [review]

**Review by loganharbour — APPROVED** [review]

**Pull request merged**